crates/wasmparser/src/validator/component.rs

```diff
@@ -0,0 +1,1800 @@
+            }
+            ComponentTypeDef::Compound(_) => {
+                // TODO: check with spec authors to see if this behavior is correct
+                return Err(BinaryReaderError::new(
+                    "imports of compound types are not supported",
+                    offset,
+                ));
+            }
+        };
+
+        check_max(len, 0, max, desc, offset)?;
+
+        if self
+            .imports
+            .insert(
+                import.name.to_string(),
+                ComponentEntityType::from_type_def(ty.clone(), "import", offset)?,
+            )
+            .is_some()
+        {
+            return Err(BinaryReaderError::new(
+                format!("duplicate import name `{}` already defined", import.name),
+                offset,
+            ));
+        }
+
+        Ok(())
+    }
+
+    pub fn add_export(&mut self, export: crate::ComponentExport, offset: usize) -> Result<()> {
+        let ty = match export.kind {
+            crate::ComponentExportKind::Module(idx) => {
+                ComponentEntityType::Module(self.module_at(idx, offset)?.clone())
+            }
+            crate::ComponentExportKind::Component(idx) => {
+                ComponentEntityType::Component(self.component_at(idx, offset)?.clone())
+            }
+            crate::ComponentExportKind::Instance(idx) => {
+                ComponentEntityType::Instance(self.component_instance_at(idx, offset)?.clone())
+            }
+            crate::ComponentExportKind::Function(idx) => {
+                ComponentEntityType::Func(self.component_function_at(idx, offset)?.clone())
+            }
+            crate::ComponentExportKind::Value(idx) => {
+                ComponentEntityType::Value(self.value_at(idx, offset)?.clone())
+            }
+            crate::ComponentExportKind::Exports(exports) => {
+                ComponentEntityType::Instance(self.instantiate_exports(exports.as_ref(), offset)?)
+            }
+        };
+
+        if self.exports.insert(export.name.to_string(), ty).is_some() {
+            return Err(BinaryReaderError::new(
+                format!("duplicate export name `{}` already defined", export.name),
+                offset,
+            ));
+        }
+
+        Ok(())
+    }
+
+    pub fn lift_function(
+        &mut self,
+        type_index: u32,
+        func_index: u32,
+        options: &[CanonicalOption],
+        offset: usize,
+    ) -> Result<()> {
+        let ty = self.types.function_type_at(type_index, offset)?;
+        let core_ty = self.core_function_at(func_index, offset)?;
+
+        if ty.core_type != *core_ty {
+            return Err(BinaryReaderError::new(
+                "lowered function type does not match core function type",
+                offset,
+            ));
+        }
+
+        check_options(options, offset)?;
+        self.functions.push(FuncIndexType::Component(ty.clone()));
+
+        Ok(())
+    }
```

Comment:
One thing I'd expect to get validated here as well which I don't believe is currently validated is the canonical ABI of the `into` instance.

I suspect though that at a spec level this is somewhat lacking. But for example something about lifting needs to validate that the function takes a string and as a result some malloc function has been specified somewhere. I'm not 100% sure what the vision for this is but it's something we'll need to validate.

Reply:
Yep, I have this on my TODO list that I'll add before we merge this PR.